src/modalities/training/gradient_clipping/fsdp_gradient_clipper.py

```diff
@@ -79,6 +88,135 @@ def clip_gradients(self) -> torch.Tensor:
+        """
+        Clips the gradients of the wrapped model using the specified maximum norm and norm type.
+
+        Returns:
+            torch.Tensor: The gradient norm after clipping.
+        """
+        gradient_norm_score = FSDP2GradientClipper.clip_grad_norm_(
+            parameters=self.wrapped_model.parameters(),
+            max_norm=self.max_norm,
+            norm_type=self.norm_type.value,
+            error_if_nonfinite=True,
+            foreach=True,
+        )
+        return gradient_norm_score
+
+    @staticmethod
+    def clip_grad_norm_(
+        parameters: torch.Tensor | Iterable[torch.Tensor],
+        max_norm: float,
+        norm_type: float = 2.0,
+        error_if_nonfinite: bool = False,
+        foreach: Optional[bool] = None,
+    ) -> torch.Tensor:
+        """
+        Clip the gradient norm of an iterable of parameters.
+
+        Gradient norm clipping requires computing the gradient norm over the entire model.
+        `torch.nn.utils.clip_grad_norm_` only computes gradient norm along DP/FSDP/TP dimensions.
+
+        TODO: for pipeline parallelism, we need to implement it like here:
+        https://github.com/pytorch/torchtitan/blob/b291ad662493b63d25b038a30a915082d3617baf/torchtitan/distributed/utils.py#L245
+        I removed all the code w.r.t. pipeline parallelism for now.
+
+        Args:
+            parameters: an iterable of Tensors or a single Tensor that will have gradients normalized
+            max_norm (float): max norm of the gradients
+            norm_type (float): type of the used p-norm. Can be ``'inf'`` for
+                infinity norm.
+            error_if_nonfinite (bool): if True, an error is thrown if the total
+                norm of the gradients from :attr:`parameters` is ``nan``,
+                ``inf``, or ``-inf``. Default: False (will switch to True in the future)
+            foreach (bool): use the faster foreach-based implementation.
+                If ``None``, use the foreach implementation for CUDA and CPU native tensors and silently
+                fall back to the slow implementation for other device types.
+                Default: ``None``
+
+        Returns:
+            Total norm of the parameter gradients (viewed as a single vector).
+
+        """
+        grads = [p.grad for p in parameters if p.grad is not None]
+        total_norm = torch.nn.utils.get_total_norm(
+            tensors=grads, norm_type=norm_type, error_if_nonfinite=error_if_nonfinite, foreach=foreach
+        )
+
+        # Inspired by torch titan
+        # If total_norm is a DTensor, the placements must be `torch.distributed._tensor.ops.math_ops._NormPartial`.
+        # We can simply reduce the DTensor to get the total norm in this tensor's process group
+        # and then convert it to a local tensor.
+        # NOTE: It has the purpose to return a reduced total_norm tensor whose .item() would return the correct value
+        if isinstance(total_norm, DTensor):
```

Comment:
To my understanding, the current approach does not work for PP. If so, it would be good to raise an Exception when PP > 1.

Reply:
Torch titan passes in the pp_mesh. We don't have this anyways yet and throw an exception/raise a warning if someone sets pp > 1 in the device mesh